### ISSUE TITLE
Dealing with 'im' vs 'vol' functions shouldn't justify separate method implementations

### ISSUE THREAD

**Issue by vineetbansal** [vineetbansal/Cryo#20]
FBBasis and HBBasis classes differ in the `expand_t` implementations because they deal with `im_to_vec/vec_to_im` vs. `vol_to_vec/vec_to_vol` functions internally. We can still keep the `im_to_vec/vol_to_vec` functions since they're commonly used, but also provide a `vec_to_hypercube(ndim)` and `hypercube_to_vec(ndim)` etc. function so that implementation of common methods don't have to be duplicated in subclasses.

There might be other examples of this in the code.

**Comment by janden** [comment]
I think this makes sense to combine these functions, although I'm not sure about `hypercube` since it evokes a sense of four dimensions or higher. Maybe `array_to_vec`, `ndarray_to_vec`, or just `to_vec`/`from_vec` or `vectorize`/`unvectorize`?

Also, regarding the original comment, `expand_t` should be very different in 2D and 3D. I think when you had the (precalculated) matrix multiplication in there, that wasn't true, but it should be the case if you use the original implementations from the MATLAB version.

**Comment by vineetbansal** [comment]
Yes .. these names make more sense than `hypercube`. Once we have an `Image` class and a `Volume` class (both types of `NDArray`), these methods can be added to the base class.

For the `expand_t` methods (not the precalculated matrix version), I am seeing differences only in the fact that one is dealing with 2d, the other with 3d, so they're identical for all purposes except for a single conceptual no. - the no. of dimensions:

![expand_t_diffs](https://user-images.githubusercontent.com/318858/54538210-3d944900-496a-11e9-99b1-ffd6a4ccdcc8.png)

**Comment by janden** [comment]
> Yes .. these names make more sense than `hypercube`. Once we have an `Image` class and a `Volume` class (both types of `NDArray`), these methods can be added to the base class.

Ok. That makes sense.

> For the `expand_t` methods (not the precalculated matrix version), I am seeing differences only in the fact that one is dealing with 2d, the other with 3d, so they're identical for all purposes except for a single conceptual no. - the no. of dimensions:

Sorry I was confused. I was thinking of the `evaluate_t` method. In fact, `expand_t` should be removed. It doesn't make sense in most settings and isn't really well defined. It was included for symmetry in the basis objects, but is never used anywhere. I had removed it from the ASPIRE code but hadn't for the Cov3D code yet.

**Comment by junchaoxia** [comment]
@janden I also saw the expand function in the Matlab version of ffb method. Could you confirm that it is really needed?

**Comment by janden** [comment]
> @janden I also saw the expand function in the Matlab version of ffb method. Could you confirm that it is really needed?

`expand` is neeed. `expand_t` is not.

**Comment by junchaoxia** [comment]
@janden From Zhizhen' paper of 2016, the coefficients of the fast FB method are obtained in a different way (NUFFT and 1DFFT), not the least squares method in the current implementation of `expand` function in the FB method. My understanding is that `expand` is the same as `evaluate_t` in the fast FB method. Do you mean that the same `expand` in FB method should be kept in the fast version?

**Comment by janden** [comment]
In both FB and FFB (fast FB), `expand` is defined as the pseudo-inverse (minimizing squared error) of `evaluate`. This is calculated using conjugate gradient which calls `evaluate` and `evaluate_t`. The FFB implementation provides faster functions `evaluate` and `evaluate_t` compared to FB, so as a result, its `expand` runs faster than the corresponding `expand` in FB.

In Jane's paper, they use `evaluate_t` instead of `expand` because it is faster, but it's not as accurate. Sometimes, this is a reasonable tradeoff, but sometimes you want the most accurate pseudo-inverse you can calculate (for example, `basis_expand(basis, basis_evaluate(basis, v))` always equals `v`, but this is not true of `basis_evaluate_t(basis, basis_evaluate(basis, v))`.

**Comment by junchaoxia** [comment]
@janden Good to know. That can not be obtained from the paper. By the way, I assume that the same idea is applicable to the fast FB 3D.

**Comment by janden** [comment]
> I assume that the same idea is applicable to the fast FB 3D.

Yes. All of the above holds for both 2D and 3D.
